turbo/src/test/kotlin/dev/hotwire/turbo/util/TurboUriHelperTest.kt

```diff
@@ -0,0 +1,79 @@
+package dev.hotwire.turbo.util
+
+import android.content.Context
+import android.net.Uri
+import android.os.Build
+import androidx.test.core.app.ApplicationProvider
+import dev.hotwire.turbo.BaseUnitTest
+import kotlinx.coroutines.Dispatchers
+import kotlinx.coroutines.ExperimentalCoroutinesApi
+import kotlinx.coroutines.test.TestCoroutineDispatcher
+import kotlinx.coroutines.test.resetMain
+import kotlinx.coroutines.test.runTest
+import kotlinx.coroutines.test.setMain
+import org.assertj.core.api.Assertions.assertThat
+import org.junit.Before
+import org.junit.Test
+import org.junit.runner.RunWith
+import org.robolectric.RobolectricTestRunner
+import org.robolectric.Shadows
+import org.robolectric.annotation.Config
+import java.io.File
+
+@ExperimentalCoroutinesApi
+@RunWith(RobolectricTestRunner::class)
+@Config(sdk = [Build.VERSION_CODES.O])
+class TurboUriHelperTest : BaseUnitTest() {
+
+    private val testDispatcher: TestCoroutineDispatcher = TestCoroutineDispatcher()
+
+    private lateinit var context: Context
+    private lateinit var turboUriHelper: TurboUriHelper
+
+    @Before
+    override fun setup() {
+        super.setup()
+        Dispatchers.setMain(testDispatcher)
+        dispatcherProvider.io = Dispatchers.Main
+
+        context = ApplicationProvider.getApplicationContext()
+        turboUriHelper = TurboUriHelper(context)
+    }
+
+    override fun teardown() {
+        super.teardown()
+        Dispatchers.resetMain()
+        testDispatcher.cleanupTestCoroutines()
+    }
+
+    @Test
+    fun validUriIsWrittenToFileSuccessfully() = runTest {
+        val inputFile = File("/tmp/file.txt")
+        val inputFileUri = Uri.fromFile(inputFile)
+        Shadows.shadowOf(context.contentResolver).registerInputStream(inputFileUri, "fileContent".byteInputStream())
+
+        val destFile = turboUriHelper.writeFileTo(inputFileUri, TurboFileProvider.directory(context))
+
+        assertThat(destFile).isNotNull()
+    }
+
+    @Test
+    fun pathTraversingUriIsFailsToWriteToFileV1() = runTest {
+        val inputFileUri = Uri.parse("../../tmp/file.txt")
+        Shadows.shadowOf(context.contentResolver).registerInputStream(inputFileUri, "fileContent".byteInputStream())
+
+        val destFile = turboUriHelper.writeFileTo(inputFileUri, TurboFileProvider.directory(context))
+
+        assertThat(destFile).isNull()
+    }
+
+    @Test
+    fun pathTraversingUriIsFailsToWriteToFileV2() = runTest {
+        val inputFileUri = Uri.parse("content://malicious.app?path=/data/data/malicious.app/files/file.txt&name=../../file.txt")
```

Comment:
Maybe use slightly clearer (even if really verbose) names for the test names? I was confused at first by the `v1` vs `v2` distinction at first, not knowing if was just to distinguish the test names or had a more significant meaning.

How about:
```kotlin
`pathTraversingUriWithRelativePathFailsToWriteToFile()` and
`pathTraversingUriWithNameArgFailsToWriteToFile()`

```